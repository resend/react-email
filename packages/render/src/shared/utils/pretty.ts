import type { Options } from 'prettier';
import html from 'prettier/plugins/html';
import { format } from 'prettier/standalone';

const defaults: Options = {
  endOfLine: 'lf',
  tabWidth: 2,
  plugins: [html],
  parser: 'html',
};

export const pretty = (str: string, options: Options = {}) => {
  return format(str, {
    ...defaults,
    ...options,
  });
};
