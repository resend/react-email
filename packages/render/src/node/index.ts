import type { Options } from '../shared/options';
import { render } from './render';

/**
 * @deprecated use {@link render}
 */
export const renderAsync = (element: React.ReactElement, options?: Options) => {
  return render(element, options);
};

export * from '../shared/options';
export * from '../shared/utils/pretty';
export * from '../shared/utils/to-plain-text';
export * from '../shared/utils/to-resend-template';
export * from './render';
