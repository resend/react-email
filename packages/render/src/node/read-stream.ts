import { Writable } from 'node:stream';
import type {
  PipeableStream,
  ReactDOMServerReadableStream,
} from 'react-dom/server.browser';

const decoder = new TextDecoder('utf-8');

export const readStream = async (
  stream: PipeableStream | ReactDOMServerReadableStream,
) => {
  let result = '';

  if ('pipeTo' in stream) {
    // means it's a readable stream
    const writableStream = new WritableStream({
      write(chunk: BufferSource) {
        result += decoder.decode(chunk);
      },
    });
    await stream.pipeTo(writableStream);
  } else {
    const writable = new Writable({
      write(chunk: BufferSource, _encoding, callback) {
        result += decoder.decode(chunk);

        callback();
      },
    });
    stream.pipe(writable);

    await new Promise<void>((resolve, reject) => {
      writable.on('error', reject);
      writable.on('close', () => {
        resolve();
      });
    });
  }

  return result;
};
