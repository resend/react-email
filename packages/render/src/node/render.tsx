import { convert } from 'html-to-text';
import { Suspense } from 'react';
import type { Options } from '../shared/options';
import { plainTextSelectors } from '../shared/plain-text-selectors';
import { pretty } from '../shared/utils/pretty';
import { readStream } from './read-stream';

export const render = async (
  element: React.ReactElement,
  options?: Options,
) => {
  const suspendedElement = <Suspense>{element}</Suspense>;
  const reactDOMServer = await import('react-dom/server');

  let html!: string;
  if (Object.hasOwn(reactDOMServer, 'renderToReadableStream')) {
    html = await readStream(
      await reactDOMServer.renderToReadableStream(suspendedElement),
    );
  } else {
    await new Promise<void>((resolve, reject) => {
      const stream = reactDOMServer.renderToPipeableStream(suspendedElement, {
        async onAllReady() {
          html = await readStream(stream);
          resolve();
        },
        onError(error) {
          reject(error as Error);
        },
      });
    });
  }

  if (options?.plainText) {
    return convert(html, {
      selectors: plainTextSelectors,
      ...options.htmlToTextOptions,
    });
  }

  const doctype =
    '<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">';

  const document = `${doctype}${html.replace(/<!DOCTYPE.*?>/, '')}`;

  if (options?.pretty) {
    return pretty(document);
  }

  return document;
};
