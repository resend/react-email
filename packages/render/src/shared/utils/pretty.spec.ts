import { promises as fs } from 'node:fs';
import path from 'node:path';
import { pretty } from './pretty';

describe('pretty', () => {
  it("should prettify Preview component's complex characters correctly", async () => {
    const stripeHTML = await fs.readFile(
      path.resolve(__dirname, './stripe-email.html'),
      'utf8',
    );

    expect(await pretty(stripeHTML)).toMatchSnapshot();
  });

  test('if mso syntax does not wrap', async () => {
    expect(
      await pretty(
        `<span><!--[if mso]><i style="mso-font-width:100%;mso-text-raise:12" hidden>&#8202;&#8202;</i><![endif]--></span>`,
      ),
    ).toMatchSnapshot();
  });
});
