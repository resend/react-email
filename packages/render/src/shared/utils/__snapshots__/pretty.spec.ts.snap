// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`pretty > if mso syntax does not wrap 1`] = `
"<span
  ><!--[if mso]><i style="mso-font-width:100%;mso-text-raise:12" hidden>&#8202;&#8202;</i><![endif]--></span
>
"
`;

exports[`pretty > should prettify Preview component's complex characters correctly 1`] = `
"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<!--$-->
<html dir="ltr" lang="en">
  <head>
    <meta content="text/html; charset=UTF-8" http-equiv="Content-Type" />
    <meta name="x-apple-disable-message-reformatting" />
  </head>
  <div
    style="display:none;overflow:hidden;line-height:1px;opacity:0;max-height:0;max-width:0">
    You&#x27;re now ready to make live transactions with Stripe!
    <div>
       ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿
    </div>
  </div>
  <body
    style='background-color:#f6f9fc;font-family:-apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,"Helvetica Neue",Ubuntu,sans-serif'>
    <table
      align="center"
      width="100%"
      border="0"
      cellpadding="0"
      cellspacing="0"
      role="presentation"
      style="max-width:37.5em;background-color:#ffffff;margin:0 auto;padding:20px 0 48px;margin-bottom:64px">
      <tbody>
        <tr style="width:100%">
          <td>
            <table
              align="center"
              width="100%"
              border="0"
              cellpadding="0"
              cellspacing="0"
              role="presentation"
              style="padding:0 48px">
              <tbody>
                <tr>
                  <td>
                    <img
                      alt="Stripe"
                      height="21"
                      src="/static/stripe-logo.png"
                      style="display:block;outline:none;border:none;text-decoration:none"
                      width="49" />
                    <hr
                      style="width:100%;border:none;border-top:1px solid #eaeaea;border-color:#e6ebf1;margin:20px 0" />
                    <p
                      style="font-size:16px;line-height:24px;margin:16px 0;color:#525f7f;text-align:left">
                      Thanks for submitting your account information.
                      You&#x27;re now ready to make live transactions with
                      Stripe!
                    </p>
                    <p
                      style="font-size:16px;line-height:24px;margin:16px 0;color:#525f7f;text-align:left">
                      You can view your payments and a variety of other
                      information about your account right from your dashboard.
                    </p>
                    <a
                      href="https://dashboard.stripe.com/login"
                      style="line-height:100%;text-decoration:none;display:block;max-width:100%;mso-padding-alt:0px;background-color:#656ee8;border-radius:5px;color:#fff;font-size:16px;font-weight:bold;text-align:center;width:100%;padding:10px 10px 10px 10px"
                      target="_blank"
                      ><span
                        ><!--[if mso]><i style="mso-font-width:500%;mso-text-raise:15" hidden>&#8202;</i><![endif]--></span
                      ><span
                        style="max-width:100%;display:inline-block;line-height:120%;mso-padding-alt:0px;mso-text-raise:7.5px"
                        >View your Stripe Dashboard</span
                      ><span
                        ><!--[if mso]><i style="mso-font-width:500%" hidden>&#8202;&#8203;</i><![endif]--></span
                      ></a
                    >
                    <hr
                      style="width:100%;border:none;border-top:1px solid #eaeaea;border-color:#e6ebf1;margin:20px 0" />
                    <p
                      style="font-size:16px;line-height:24px;margin:16px 0;color:#525f7f;text-align:left">
                      If you haven&#x27;t finished your integration, you might
                      find our<!-- -->
                      <a
                        href="https://stripe.com/docs"
                        style="color:#556cd6;text-decoration-line:none"
                        target="_blank"
                        >docs</a
                      >
                      <!-- -->handy.
                    </p>
                    <p
                      style="font-size:16px;line-height:24px;margin:16px 0;color:#525f7f;text-align:left">
                      Once you&#x27;re ready to start accepting payments,
                      you&#x27;ll just need to use your live<!-- -->
                      <a
                        href="https://dashboard.stripe.com/login?redirect=%2Fapikeys"
                        style="color:#556cd6;text-decoration-line:none"
                        target="_blank"
                        >API keys</a
                      >
                      <!-- -->instead of your test API keys. Your account can
                      simultaneously be used for both test and live requests, so
                      you can continue testing while accepting live payments.
                      Check out our<!-- -->
                      <a
                        href="https://stripe.com/docs/dashboard"
                        style="color:#556cd6;text-decoration-line:none"
                        target="_blank"
                        >tutorial about account basics</a
                      >.
                    </p>
                    <p
                      style="font-size:16px;line-height:24px;margin:16px 0;color:#525f7f;text-align:left">
                      Finally, we&#x27;ve put together a<!-- -->
                      <a
                        href="https://stripe.com/docs/checklist/website"
                        style="color:#556cd6;text-decoration-line:none"
                        target="_blank"
                        >quick checklist</a
                      >
                      <!-- -->to ensure your website conforms to card network
                      standards.
                    </p>
                    <p
                      style="font-size:16px;line-height:24px;margin:16px 0;color:#525f7f;text-align:left">
                      We&#x27;ll be here to help you with any step along the
                      way. You can find answers to most questions and get in
                      touch with us on our<!-- -->
                      <a
                        href="https://support.stripe.com/"
                        style="color:#556cd6;text-decoration-line:none"
                        target="_blank"
                        >support site</a
                      >.
                    </p>
                    <p
                      style="font-size:16px;line-height:24px;margin:16px 0;color:#525f7f;text-align:left">
                      — The Stripe team
                    </p>
                    <hr
                      style="width:100%;border:none;border-top:1px solid #eaeaea;border-color:#e6ebf1;margin:20px 0" />
                    <p
                      style="font-size:12px;line-height:16px;margin:16px 0;color:#8898aa">
                      Stripe, 354 Oyster Point Blvd, South San Francisco, CA
                      94080
                    </p>
                  </td>
                </tr>
              </tbody>
            </table>
          </td>
        </tr>
      </tbody>
    </table>
  </body>
</html>
<!--/$-->
"
`;
